Warn on case mismatch while opening a file on macOS

Kanabenki opened the pull request:
Fixes #23441

Kanabenki committed:
Warn on case mismatch on macOS

aaronfranke commented:
In my opinion this should happen on Linux too, so that projects can be cross-compatible. Godot should not allow names that only differ by case, and require names to have the same casing everywhere.

Kanabenki commented:
It is true that it could be possible to have the same problem on Linux (for example for a project stored on a USB key with a FAT32 filesystem).

Kanabenki committed:
Fix wrong warning on case sensitive filesystem

Do not issue a warning on a case sensitive fs if two files have the same
case-insensitive name

Kanabenki commented:
So, should I remove the macos check so that the warning can be issued on any unix system ?

aaronfranke commented:
I would appreciate feedback from the Godot devs about this one.

reduz commented on drivers/unix/file_access_unix.cpp:
```diff
@@ -104,6 +108,26 @@ Error FileAccessUnix::_open(const String &p_path, int p_mode_flags) {
+		struct dirent *dir_entry;
+		if (dir) {
+			String base_file = path.get_file();
+			while ((dir_entry = readdir(dir)) != NULL) {
```
This is very inefficient, please look for a different way to do it. Maybe the stat() function is a better fit for this?

Kanabenki replied:
The problem is that stat doesn't return the name of the requested file, so there is no way to check the case. I also didn't find a function similar to FindFirstFileW on windows. I'll look into another solution.

Kanabenki commented:
Thanks for the proposal ! I'm currently stuck trying to find a suitable way of checking for the problem that is compatible with both linux and macos (and haven't had much time to work on it recently), so I would gladly accept help on this one.

EmrysMyrddin commented:
I've found a way to get the real path on disk of a file. My problem is that it also resolves symlinks and relatives paths (it resolves `.` and `..`)...

I'm a total noob on Godot, but I imagine that including a file can be donne be providing relative links ?

EmrysMyrddin commented:
Since I dont' have the right to push on your branch and that I haven't used any of your code, I have open a new PR (#24860) to avoid messing with git history.

akien-mga commented:
Closing as superseded by #24860 which looks like a better option. Still, thanks for the contribution which help get closer to a good solution!
